### ISSUE TITLE
Add optional more specific MIME type for versioning

### ISSUE THREAD

**Issue by nichtich** [gbv/daia#23]
Does DAIA need to be versioned? See http://blog.steveklabnik.com/posts/2011-07-03-nobody-understands-rest-or-http#i_want_my_api_to_be_versioned for a solution.

**Comment by lahmann** [comment]
This would be a reasonable solution for keeping DAIA open for breaking changes in its data structure in the future. We would need to define a DAIA-json Mime-type and by default the server should return the latest version implemented in its service if no specific version is given in the Accept-Header.
If only unsupported Mime-type versions are given in the Accept-Header the server should return a 406 (maybe with a list of supported Mime-types?).

**Comment by nichtich** [comment]
Selecting a DAIA version from client may be supported by a custom MIME type, e.g.

```
Accept: application/vnd.daia.v1+json
```

But I hesitate to require this MIME type to be returned by all servers. [GitHub API](https://developer.github.com/v3/media/) uses a custom response header `X-GitHub-Media-Type` and the DAIA Draft 0.9.8 contains a custom `X-DAIA-Version` header, so we could do

```
Content-type: application/json
X-DAIA-Media-Type: application/vnd.daia.v1+json
```

or 

```
Content-type: application/json
X-DAIA-Version: 1.0.0
```

I still prefer the second because it gives a more precise version number and explicitly.

How should a DAIA 1.0 server react if receiving a `application/vnd.daia.v2+json` request?

**Comment by nichtich** [comment]
I'd also make response header `X-DAIA-Version` mandatory (it's a SHOULD in version 0.9.8).

**Comment by lahmann** [comment]
Selecting a DAIA version from client side via custom Mime type could also be achieved by using additional parameters:

```
Accept: application/vnd.daia+json; version=1.0.0
```

or

```
X-DAIA-Media-Type: application/vnd.daia+json; version=1.0.0
```

But, the suggested headers

```
Content-type: application/json
X-DAIA-Version: 1.0.0
```

are in deed favourable as they are simple (keeping customization minimal) and explicit. Although I think using `X-DAIA-Media-Version` would be even more explicit as `X-DAIA-Version` seems to imply that the request should be processed by the DAIA service version 1.0.0 whereas we are only requesting a DAIA document in the format of DAIA version 1.0.0.
In other words, dealing with other versions of the service (which could mean new capabilities) is a complete different issue and should not be addressed by the Accept headers.

But maybe that's just me being nit-pickish.

**Comment by nichtich** [comment]
I am not sure about the use of of selecting a DAIA version from client:
- If a client requests DAIA 1.0.0 at an old DAIA server (DAIA 0.5), the version is ignored anyway
- If a client requests DAIA 1.x.x at a DAIA 1.y.y server, everything is fine anyway. The server might know which optional features the client does not understand (x > y) but the client can ignore these features anyway  
- If a client requests DAIA 1.x.x at a DAIA 2.x.x server, an error should be returned

So the only use case is a possible future DAIA 2.x.x breaking backwards-compatibility. Is this reason enough to introduce the inclusion of DAIA version in `Accept` header and requirement to check the version by server?

**Comment by lahmann** [comment]
> I am not sure about the use of of selecting a DAIA version from client

That's what I meant - therefore `X-DAIA-Media-Version` is more appropriate as the client should request the response media type explicitly but not the service version (and thus implicitly the response media type). In other words, I propose a service version agnostic approach where the `Accept` header defines only the document version expected - therefore the better naming of the custom `Accept` header would be (imo) `X-DAIA-Media-Version`.

However, if you have strong objections I can also live with `X-DAIA-Version`. :)

**Comment by nichtich** [comment]
The name does not matter much but the specification must tell what DAIA a servers is expected to do when receiving a `X-DAIA-Media-Version/Accept/X-DAIA-Version` header. As I wrote 4 hours ago, the header will only be relevant when there will be an incompatible future DAIA version 2.0 **and** if a DAIA 2.0 server should be accessible at the same base URL as 1.0 . Either we assume that this is not important or all DAIA 1.0 servers are required to check the `X-DAIA-Media-Version/Accept/X-DAIA-Version` header.

**Comment by nichtich** [comment]
I decided to not include such header and more rules about DAIA versioning. It would only add complexity for little benefit.
